src/v/kafka/server/handlers/describe_producers.cc

```diff
@@ -0,0 +1,177 @@
+// by the Apache License, Version 2.0
+
+#include "kafka/server/handlers/describe_producers.h"
+
+#include "cluster/errc.h"
+#include "cluster/partition_manager.h"
+#include "cluster/shard_table.h"
+#include "kafka/protocol/describe_producers.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/describe_producers_response.h"
+#include "kafka/server/handlers/details/security.h"
+#include "kafka/server/request_context.h"
+#include "kafka/server/response.h"
+#include "model/fundamental.h"
+#include "model/namespace.h"
+#include "resource_mgmt/io_priority.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/when_all.hh>
+
+namespace kafka {
+
+namespace {
+
+int64_t get_time_since_epoch(ss::lowres_clock::time_point last_update_time) {
+    auto diff = ss::lowres_clock::now() - last_update_time;
+    return std::chrono::system_clock::now().time_since_epoch().count()
+           - diff.count();
+}
+
+using ret_res = checked<cluster::rm_stm::transaction_set, kafka::error_code>;
+
+ss::future<checked<cluster::rm_stm::transaction_set, kafka::error_code>>
+get_producers_for_partition(request_context& ctx, model::ntp ntp) {
+    auto shard = ctx.shards().shard_for(ntp);
+    if (!shard) {
+        vlog(
+          klog.debug,
+          "Can not find shard for ntp({}) to get info about producers",
+          ntp);
+        return ss::make_ready_future<ret_res>(
+          kafka::error_code::not_leader_for_partition);
+    }
+
+    return ctx.partition_manager().invoke_on(
+      *shard, [ntp](cluster::partition_manager& pm) -> ss::future<ret_res> {
+          auto partition = pm.get(ntp);
+          if (!partition) {
+              return ss::make_ready_future<ret_res>(
+                kafka::error_code::unknown_topic_or_partition);
+          }
+
+          if (!partition->is_leader()) {
+              return ss::make_ready_future<ret_res>(
+                kafka::error_code::not_leader_for_partition);
+          }
+
+          auto rm_stm_ptr = partition->rm_stm();
+
+          if (!rm_stm_ptr) {
+              vlog(
+                klog.error,
+                "Can not get rm_stm for ntp({}). Looks like transaction are "
+                "disabled",
+                ntp);
+
+              return ss::make_ready_future<ret_res>(
+                kafka::error_code::unknown_server_error);
+          }
+
+          return rm_stm_ptr->get_transactions().then(
+            [ntp](auto tx) -> ss::future<ret_res> {
+                if (!tx.has_value()) {
+                    if (tx.error() == cluster::errc::not_leader) {
+                        return ss::make_ready_future<ret_res>(
+                          kafka::error_code::not_leader_for_partition);
+                    }
+
+                    vlog(
+                      klog.error,
+                      "get_transactions was timeouted. It means node is not "
+                      "leader "
+                      "for ntp({}) anymore",
+                      ntp);
```

Comment:
This log entry is very cryptic. Why timeout suggest that node is not the leader ? Also if it is a case why aren't we returning timeout error but unknown server error ?

Reply:
as I understand the only error `get_transactions()` may return is `cluster::errc::not_leader` so current branch is a dead code / safety net for accidental changes in the future.

I would have logged something like

```
vlog(klog.warn, "get_transactions returned unexpected error: {}", tx.error());
```